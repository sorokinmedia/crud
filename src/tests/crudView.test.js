import { configure, mount, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16/build/index'
import React from 'react'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import jsdom from 'jsdom'
import CrudView from '../components/crud/view/crudView'
import thunk from 'redux-thunk'
import { shallowToJson } from 'enzyme-to-json'
import { BrowserRouter } from 'react-router-dom'
import responseMock from './responseMock'

const { JSDOM } = jsdom;
const doc = new JSDOM('<!doctype html><html><body></body></html>');
global.document = doc;
global.window = doc.defaultView;
global.navigator = {
	userAgent: 'node.js',
};

// const {SignUpClass} = require('../../../containers/Page/signup');

configure({adapter: new Adapter()});

const mockStore = configureStore([thunk]);

describe('it should render CrudView', () => {
	let container, modelName = 'model';

	beforeAll(() => {
		container = shallow(<CrudView handleSubmit={jest.fn} store={mockStore({ crudModels: { [modelName]: responseMock }, crudFilterValues: {} })} modelName={modelName} />);
	});

	it('it should render SignUp connected container', () => {
		expect(shallowToJson(container)).toMatchSnapshot();
	});
});

describe('should mount CrudView', () => {
	let container, store;
	const modelName = 'model';
	const openUpdateFrom = jest.fn();
	const handleDelete = jest.fn();
	const handleRestore = jest.fn();
	const crudActionsFunc = jest.fn((action, elem) => {
		switch (action.id) {
		case 'update':
			openUpdateFrom(action, elem);
			break;
		case 'delete':
			handleDelete(action, elem);
			break;
		case 'restore':
			handleRestore(action, elem);
			break;
		default:
			return null
		}
	});
	const storeData = {
		crudActionsFunc: { [modelName]: crudActionsFunc },
		crudModels: { [modelName]: responseMock },
		crudFilterValues: { [modelName]: {} },
		crudColumns: { [modelName]: responseMock.data.columns.map(e => ({ ...e, visible: true })) }
	};

	beforeAll(() => {
		store = mockStore(storeData);
		container = mount(<Provider store={store}>
			<BrowserRouter>
				<CrudView
					crudRead="url"
					modelName={modelName}
				/>
			</BrowserRouter>
		</Provider>);
	});

	it('should invoke delete func', () => {
		const button = container.find('.anticon-delete').first();

		button.simulate('click');

		expect(handleDelete).toHaveBeenCalled();
	});

	it('should invoke change func', () => {
		const button = container.find('.anticon-edit').first();

		button.simulate('click');

		expect(openUpdateFrom).toHaveBeenCalled();
	});

	it('should invoke restore func', () => {
		const button = container.find('.anticon-select').first();

		button.simulate('click');

		expect(handleRestore).toHaveBeenCalled();
	});

	it('should render items', () => {
		expect(container.find('.ant-table-scroll .ant-table-tbody tr').length)
			.toBe(storeData.crudModels[modelName].data.items.length);
	});

	it('should render columns', () => {
		expect(container.find('.ant-table-scroll th.crud-table-column').length)
			.toBe(storeData.crudModels[modelName].data.columns.length);
	});
});
