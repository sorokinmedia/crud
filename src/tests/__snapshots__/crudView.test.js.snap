// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render CrudView it should render SignUp connected container 1`] = `
<CrudView
  fetchCrudChildren={[Function]}
  fetchCrudModels={[Function]}
  fixActionColumn={true}
  handleSubmit={[Function]}
  items={
    Object {
      "data": Object {
        "columns": Array [
          Object {
            "filter": Object {
              "can": true,
              "defaultValue": "",
              "query": "",
              "type": "input_number",
            },
            "id": "id",
            "order": Object {
              "can": true,
              "orders": Array [
                "SORT_ASC",
                "SORT_DESC",
              ],
            },
            "title": "ID",
            "type": "integer",
          },
          Object {
            "filter": Object {
              "can": true,
              "defaultValue": "",
              "query": "",
              "type": "input_text",
            },
            "id": "name",
            "order": Object {
              "can": true,
              "orders": Array [
                "SORT_ASC",
                "SORT_DESC",
              ],
            },
            "title": "Название",
            "type": "string",
          },
          Object {
            "filter": Object {
              "can": true,
              "defaultValue": "",
              "query": "",
              "type": "boolean",
            },
            "id": "is_deleted",
            "order": Object {
              "can": true,
              "orders": Array [
                "SORT_ASC",
                "SORT_DESC",
              ],
            },
            "title": "Удален",
            "type": "boolean",
          },
          Object {
            "filter": Object {
              "can": false,
              "defaultValue": "",
              "query": "",
              "type": "",
            },
            "id": "actions",
            "order": Object {
              "can": false,
              "orders": Array [],
            },
            "title": "Действия",
            "type": "actions",
          },
        ],
        "count": 5,
        "filter": Object {
          "id": null,
          "is_deleted": null,
          "limit": 100,
          "name": null,
          "offset": 0,
          "order": "SORT_DESC",
          "order_by": "id",
        },
        "items": Array [
          Object {
            "actions": Array [
              Object {
                "icon": "fa fa-pencil-square-o",
                "id": "update",
                "method": "POST",
                "name": "Изменить",
                "type": "query",
                "url": "/v1/admin/object/property/type/5/update",
              },
              Object {
                "icon": "fa fa-reply",
                "id": "restore",
                "method": "POST",
                "name": "Восстановить",
                "type": "query",
                "url": "/v1/admin/object/type/3/restore",
              },
              Object {
                "icon": "fa fa-trash",
                "id": "delete",
                "method": "POST",
                "name": "Удалить",
                "type": "query",
                "url": "/v1/admin/object/property/type/5/delete",
              },
            ],
            "id": 5,
            "is_deleted": "Нет",
            "name": "Иное",
          },
          Object {
            "actions": Array [
              Object {
                "icon": "fa fa-pencil-square-o",
                "id": "update",
                "method": "POST",
                "name": "Изменить",
                "type": "query",
                "url": "/v1/admin/object/property/type/4/update",
              },
              Object {
                "icon": "fa fa-trash",
                "id": "delete",
                "method": "POST",
                "name": "Удалить",
                "type": "query",
                "url": "/v1/admin/object/property/type/4/delete",
              },
            ],
            "id": 4,
            "is_deleted": "Нет",
            "name": "Сантехника",
          },
          Object {
            "actions": Array [
              Object {
                "icon": "fa fa-pencil-square-o",
                "id": "update",
                "method": "POST",
                "name": "Изменить",
                "type": "query",
                "url": "/v1/admin/object/property/type/3/update",
              },
              Object {
                "icon": "fa fa-trash",
                "id": "delete",
                "method": "POST",
                "name": "Удалить",
                "type": "query",
                "url": "/v1/admin/object/property/type/3/delete",
              },
            ],
            "id": 3,
            "is_deleted": "Нет",
            "name": "Ремонт",
          },
          Object {
            "actions": Array [
              Object {
                "icon": "fa fa-pencil-square-o",
                "id": "update",
                "method": "POST",
                "name": "Изменить",
                "type": "query",
                "url": "/v1/admin/object/property/type/2/update",
              },
              Object {
                "icon": "fa fa-trash",
                "id": "delete",
                "method": "POST",
                "name": "Удалить",
                "type": "query",
                "url": "/v1/admin/object/property/type/2/delete",
              },
            ],
            "id": 2,
            "is_deleted": "Нет",
            "name": "Мебель",
          },
          Object {
            "actions": Array [
              Object {
                "icon": "fa fa-pencil-square-o",
                "id": "update",
                "method": "POST",
                "name": "Изменить",
                "type": "query",
                "url": "/v1/admin/object/property/type/1/update",
              },
              Object {
                "icon": "fa fa-trash",
                "id": "delete",
                "method": "POST",
                "name": "Удалить",
                "type": "query",
                "url": "/v1/admin/object/property/type/1/delete",
              },
            ],
            "id": 1,
            "is_deleted": "Нет",
            "name": "Техника",
          },
        ],
      },
      "messages": Array [
        Object {
          "message": "Данные получены",
          "targetField": null,
          "type": 3,
        },
      ],
      "status": 0,
    }
  }
  modelName="model"
  pageSize={20}
  scrollX={1300}
  setCrudParams={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  tableProps={Object {}}
/>
`;
